sdks/python/apache_beam/io/filesystemio.py

```diff
@@ -0,0 +1,492 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""Utilities for ``FileSystem`` implementations."""
+
+import abc
+import cStringIO
+import logging
+import multiprocessing
+import os
+import traceback
+
+__all__ = ['Downloader', 'Uploader', 'BufferedReader', 'BufferedWriter']
+
+
+class Downloader(object):
+  """A download interface for a single file.
+
+  Ranges to download are requested in get_range. Data is streamed to a
+  user-provided download_stream.
+  """
+
+  __metaclass__ = abc.ABCMeta
+
+  @abc.abstractproperty
+  def size(self):
+    """Size of the file to download"""
+
+  @abc.abstractmethod
+  def start(self, download_stream, buffer_size):
+    """Initialize downloader.
+
+    Args:
+      download_stream: (cStringIO.StringIO) A buffer where downloaded data is
+        streamed to.
+      buffer_size: Maximum range size for get_range calls.
```

Comment:
<!--new_thread; commit:970afe0523d6d5673489a70e801be6553e0ffc33; resolved:0-->
It looks like the buffer_size argument here is mostly an internal implementation detail of the specific downloader so maybe we don't need this argument?  It may be helpful for us to expose the maximum chunk size allowable for a single call to `get_range`, which we use to upper-bound the `buffer_size` parameter of `BufferedReader`.

Reply:
Added max_range_size property.

Reply:
Actually, this was removed. You can pass buffer_size= to io.BufferedReader/Writer.
See examples in gcsio.py and the unit tests.